### PR TITLE
feat!: Node.js 14->16

### LINKED ISSUE #19304: Drop support for Node.js v14
- [x] Bump tested versions, so we test 16, 18, 20
	- [x] https://github.com/prisma/ecosystem-tests/pull/3462
	- [x] https://github.com/prisma/prisma/pull/19798
	- [x] https://github.com/prisma/engines-wrapper/pull/449
	- [x] https://github.com/prisma/checkpoint-client/pull/213
	- [x] https://github.com/prisma/mini-proxy/pull/46
- [ ] Update docs (https://www.prisma.io/docs/reference/system-requirements)
	- [ ] https://github.com/prisma/docs/pull/4928 
- [x] Update CLI https://github.com/prisma/prisma/blob/d43cfff9e9135b2a2c641dede4efa4ebf1914461/packages/cli/scripts/preinstall.js#L47
- [ ] examples https://github.com/prisma/prisma-examples/pull/4750
- [x] Release notes draft
- [ ] Anything else?

### PR DESCRIPTION
**Pull request by Jolg42**
Part of https://github.com/prisma/prisma/issues/19304

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "triggerEmptyDevReleaseByIncrementingThisNumber": 0,
   "license": "Apache-2.0",
   "engines": {
-    "node": ">=14.17",
+    "node": ">=18",
```
Comment (SevInf) [inline package.json:7]: 16? Or is there a reason why we want to have 18 at the root of the monorepo?

File: .github/workflows/test-template.yml [hunk 2]
```diff
@@ -121,7 +121,7 @@ jobs:
         shard: ['1/5', '2/5', '3/5', '4/5', '5/5']
         queryEngine: ${{ fromJson(needs.determineBuildMatrixParams.outputs.queryEngine) }}
         engineProtocol: ['graphql', 'json']
-        node: [16, 18]
+        node: [16, 18, 20]
```
Comment (SevInf) [inline .github/workflows/test-template.yml:124]: We might not need 16 here if we are testing it builtkite. At least that was the reason why 14 is not there AFAIK.

File: packages/client/src/runtime/utils/serializeRawParameters.ts [hunk 3]
```diff
@@ -48,6 +48,7 @@ function encodeParameter(parameter: any, objectSerialization: 'fast' | 'slow'):
   if (isArrayBufferLike(parameter) || ArrayBuffer.isView(parameter)) {
     return {
       prisma__type: 'bytes',
+      // TODO Can we change this now in v16?
```
Comment (SevInf) [inline packages/client/src/runtime/utils/serializeRawParameters.ts:51]: Does not seem so, `ArrayBufferView` is still not accepted in `from`. Maybe we should send a PR.

File: .github/workflows/test-template.yml (hunk 2 above)
Comment (Jolg42) [reply .github/workflows/test-template.yml:124]: yes that's what we used to do indeed, I'll remove 16 here

**Comment by Jolg42** [comment]
Note: I added `- NODE_OPTIONS="--max-old-space-size=8096"` to Buildkite to unblock this PR.
It was failing with 
```
[2023-07-03T14:56:50Z] <--- Last few GCs --->
[2023-07-03T14:56:50Z] 
[2023-07-03T14:56:50Z] [30111:0x4f652d0]    35767 ms: Mark-sweep (reduce) 2045.6 (2083.3) -> 2044.8 (2083.6) MB, 882.4 / 0.0 ms  (average mu = 0.105, current mu = 0.007) allocation failure scavenge might not succeed
[2023-07-03T14:56:50Z] 
[2023-07-03T14:56:50Z] 
[2023-07-03T14:56:50Z] <--- JS stacktrace --->
[2023-07-03T14:56:50Z] 
[2023-07-03T14:56:50Z] FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
[2023-07-03T14:56:50Z]  1: 0xb090e0 node::Abort() [node]
[2023-07-03T14:56:50Z]  2: 0xa1b70e  [node]
[2023-07-03T14:56:50Z]  3: 0xce1a20 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
[2023-07-03T14:56:50Z]  4: 0xce1dc7 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
[2023-07-03T14:56:50Z]  5: 0xe99435  [node]
[2023-07-03T14:56:50Z]  6: 0xe99f16  [node]
[2023-07-03T14:56:50Z]  7: 0xea843e  [node]
[2023-07-03T14:56:50Z]  8: 0xea8e80 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
[2023-07-03T14:56:50Z]  9: 0xeabdfe v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
[2023-07-03T14:56:50Z] 10: 0xe6d33a v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [node]
[2023-07-03T14:56:50Z] 11: 0x11e61e6 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [node]
[2023-07-03T14:56:50Z] 12: 0x15d9e59  [node]
[2023-07-03T14:56:54Z] /app/node_modules/.pnpm/execa@5.1.1/node_modules/execa/lib/error.js:60
[2023-07-03T14:56:54Z] 		error = new Error(message);
[2023-07-03T14:56:54Z] 		        ^
[2023-07-03T14:56:54Z] 
[2023-07-03T14:56:54Z] Error: Command was killed with SIGABRT (Aborted): node_modules/.bin/jest --config tests/functional/jest.config.js -- typescript
```


[2023-07-03T14:56:14Z] Test Suites: 4 skipped, 161 passed, 161 of 165 total
--


[2023-07-03T14:56:54Z]
 Error: Command was killed with SIGABRT (Aborted): 
node_modules/.bin/jest --config tests/functional/jest.config.js -- 
typescript
```

File: packages/client/src/runtime/core/engines/data-proxy/utils/request.ts [hunk 4]
```diff
@@ -87,7 +87,16 @@ function buildOptions(options: RequestOptions): Https.RequestOptions {
...
     text: () => Promise.resolve(Buffer.concat(incomingData).toString()),
-    json: () => Promise.resolve(JSON.parse(Buffer.concat(incomingData).toString())),
+    json: () => {
+      const dataAsString = Buffer.concat(incomingData).toString()
+      let dataAsJson
+      try {
+        dataAsJson = JSON.parse(dataAsString)
+        return Promise.resolve(dataAsJson)
+      } catch (e) {
+        return Promise.reject(`Failed to parse JSON: ${e.message} - data.length=${dataAsString.length}`)
```
Comment (millsp) [inline packages/client/src/runtime/core/engines/data-proxy/utils/request.ts:97]: Why did you need to change this line of code?

Comment (Jolg42) [reply packages/client/src/runtime/core/engines/data-proxy/utils/request.ts:97]: To understand what's going on with the unexpected "Unexpected end of JSON input" errors.
It turns out the data is empty when that happens.

File: packages/client/helpers/functional-test/run-tests.ts [hunk 5]
```diff
@@ -71,6 +72,7 @@ async function main(): Promise<number | void> {
   }
 
   if (args['--data-proxy']) {
+    jestArgs.push('--verbose')
```
Comment (Jolg42) [inline packages/client/helpers/functional-test/run-tests.ts:75]: We could remove this now (and not wait for tests to pass to merge)

**Review by SevInf — APPROVED** [review]

File: package.json [hunk 6]
```diff
@@ -4,7 +4,7 @@
   "triggerEmptyDevReleaseByIncrementingThisNumber": 0,
   "license": "Apache-2.0",
   "engines": {
-    "node": ">=14.17",
+    "node": ">=16",
```
Comment (aqrln) [inline package.json:7]: should this be `16.13.0`?

Comment (Jolg42) [reply package.json:7]: It's not really important here it's only for us in dev but we could, I was going to check .13 in a follow up PR, I'll update these as well

Comment (aqrln) [reply package.json:7]: I think the important one is in the postinstall hook

Comment (Jolg42) [reply package.json:7]: See https://github.com/prisma/prisma/pull/20148
